src/audio/pipeline.c

```diff
@@ -532,6 +532,33 @@ int pipeline_trigger(struct pipeline *p, struct comp_dev *host, int cmd)
 
 	trace_pipe_with_ids(p, "pipeline_trigger()");
 
+	/* handle pipeline global checks before going into each components */
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+		/* if in xrun, prepare before trigger start needed */
+		if (p->xrun_bytes && p->status == COMP_STATE_PAUSED) {
+			trace_pipe_with_ids(p, "in xrun, prepare it first");
+			ret = pipeline_trigger(p, host, COMP_TRIGGER_PREPARE);
```

Comment:
pipeline_prepare is called in pipeline_xrun_recover, so why do we need additional prepare here?

Reply:
The pipeline_xrun_recover() may be empty once NO_XRUN_RECOVERY is defined, that means there will be trigger stop and trigger start from host side(no prepare trigger). That's why we need this.

Reply:
So it means that rather than empty pipeline_xrun_recover() should at least do prepare call, no matter if its disabled or not.

Reply:
> So it means that rather than empty pipeline_xrun_recover() should at least do prepare call, no matter if its disabled or not. And also you are just triggering prepare here instead of calling 

yes, that's another choice. I did it in my way as some driver might send prepare IPC in future before trigger start(ALSA do has that at this senario), but ours doesn't at the moment(we haven't added prepare IPC yet actually, you can see in ipc_stream_pcm_params() we only prepare at hw_params IPC at the moment), and this prepare is actually part of the recovery, so I want to add it here to increase the robustness of our FW(to handle with and without prepare IPC before trigger start for recovery).

> pipeline_prepare. Is this also intentional?

it was not intentional here, let me change this.